### PR TITLE
[Exam] Allow to submit early

### PR DESCRIPTION
**Pull request by sascha11110**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
We want to provide the possibility to hand in an exam early.

### Description
- [x] I've added a button to abort the exam
- [x] I've added a back/continue button with a corresponding description to continue after the student accidentally clicked *Abort*
- [x] I've ensured that all participation related components are not shown after the student submitted

### Steps for Testing
1. Log in to Artemis
2. Participate in an exam
3. Click the *Abort* button
4. Click the *Continue* button
5. Refresh and check that you can still participate
6. Click the *Abort* button
7. Finish the exam
8. Refresh and check that you cannot participate anymore (summary page is shown)

### Screenshots
Ohne-Titel

**Responsive behavior of the exam navigation bar:**
responsive

### CODE REVIEW

**Comment by jpbernius** [comment]
```
[Error] Failed to load resource: the server responded with a status of 404 () (conduction, line 0)
[Error] ERROR – Error: Uncaught (in promise): TypeError: undefined is not an object (evaluating 'this.studentExam.submitted')
isOver@https://vmbruegge60.in.tum.de/app/9.ece63cacfc5dbdd03395.chunk.js:2:1091313
canDeactivate@https://vmbruegge60.in.tum.de/app/9.ece63cacfc5dbdd03395.chunk.js:2:871
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:383569
map@[native code]
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:383471
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:383723
_tryNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796703
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796622
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:4152992
_trySubscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1466172
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465983
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
a@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1490705
_innerSub@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796882
_tryNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796784
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796622
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:732324
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1513464
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
notifyNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1639840
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1905047
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:732324
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:4152992
_trySubscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1466172
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465983
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
a@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1490705
_innerSub@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1639469
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1639306
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
notifyNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1639840
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1905047
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1513464
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1513464
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
notifyNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1797087
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1905047
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
Error: Uncaught (in promise): TypeError: undefined is not an object (evaluating 'this.studentExam.submitted')
isOver@https://vmbruegge60.in.tum.de/app/9.ece63cacfc5dbdd03395.chunk.js:2:1091313
canDeactivate@https://vmbruegge60.in.tum.de/app/9.ece63cacfc5dbdd03395.chunk.js:2:871
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:383569
map@[native code]
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:383471
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:383723
_tryNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796703
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796622
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:4152992
_trySubscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1466172
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465983
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
a@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1490705
_innerSub@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796882
_tryNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796784
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1796622
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:732324
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1513464
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
notifyNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1639840
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1905047
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:732324
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:4152992
_trySubscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1466172
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465983
subscribe@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1465845
a@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1490705
_innerSub@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1639469
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1639306
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
notifyNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1639840
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1905047
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1513464
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1513464
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305
notifyNext@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1797087
_next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1905047
next@https://vmbruegge60.in.tum.de/app/main.ece63cacfc5dbdd03395.bundle.js:2:1462305M — zone.js:832M — zone.js:784(anonymous function) — zone.js:894(anonymous function) — zone.js:421onInvokeTask — core.js:27137(anonymous function) — zone.js:420(anonymous function) — zone.js:188_ — zone.js:601(anonymous function) — zone.js:507m — zone.js:1671p — zone.js:1697
	(anonymous function) (main.ece63cacfc5dbdd03395.bundle.js:2:11101814)
	So (main.ece63cacfc5dbdd03395.bundle.js:2:35075)
	handleError (main.ece63cacfc5dbdd03395.bundle.js:2:35246)
	handleError (main.ece63cacfc5dbdd03395.bundle.js:2:11120015)
	next (main.ece63cacfc5dbdd03395.bundle.js:2:203018)
	(anonymous function) (main.ece63cacfc5dbdd03395.bundle.js:2:171605)
	__tryOrUnsub (main.ece63cacfc5dbdd03395.bundle.js:2:1464139)
	next (main.ece63cacfc5dbdd03395.bundle.js:2:1463368)
	_next (main.ece63cacfc5dbdd03395.bundle.js:2:1462531)
	next (main.ece63cacfc5dbdd03395.bundle.js:2:1462306)
	next (main.ece63cacfc5dbdd03395.bundle.js:2:1471769)
	emit (main.ece63cacfc5dbdd03395.bundle.js:2:171476)
	(anonymous function) (main.ece63cacfc5dbdd03395.bundle.js:2:10160269)
	(anonymous function) (main.ece63cacfc5dbdd03395.bundle.js:2:10155366)
	onHandleError (main.ece63cacfc5dbdd03395.bundle.js:2:196024)
	(anonymous function) (main.ece63cacfc5dbdd03395.bundle.js:2:10160375)
	(anonymous function) (main.ece63cacfc5dbdd03395.bundle.js:2:10155584)
	t (main.ece63cacfc5dbdd03395.bundle.js:2:10165208)
	(anonymous function) (main.ece63cacfc5dbdd03395.bundle.js:2:10165356)
	_ (main.ece63cacfc5dbdd03395.bundle.js:2:10163362)
	(anonymous function) (main.ece63cacfc5dbdd03395.bundle.js:2:10162134)
	m (main.ece63cacfc5dbdd03395.bundle.js:2:10175259)
	p (main.ece63cacfc5dbdd03395.bundle.js:2:10175499)
```

I do get this error on TS3 when navigating into an exam (which I am not registered for) as instructor (artemis_test_user_1) and then try to navigate back to the course overview:
https://vmbruegge60.in.tum.de/#/courses/708/exams/74 -> navigate out of exan.

**Comment by sascha11110** [comment]
@jpbernius Good catch! I've fixed it here: https://github.com/ls1intum/Artemis/pull/1884/commits/6ff4557e8d8ac679fee6eabebb03173ea577ad4e

**Comment by jpbernius** [comment]
I can reproduce this with quiz options as well.

File: src/main/webapp/app/exam/participate/exam-participation.component.ts [hunk 1]
```diff
@@ -286,7 +286,7 @@ export class ExamParticipationComponent implements OnInit, OnDestroy, ComponentC
             // if this was calculated to true by the server, we can be sure the student exam has finished
             return true;
         }
-        if (this.handInEarly || this.studentExam.submitted) {
+        if (this.handInEarly || (this.studentExam && this.studentExam.submitted)) {
```
Comment (jpbernius) [inline src/main/webapp/app/exam/participate/exam-participation.component.ts:289]: Would this lead to the same result? I am unsure.
```suggestion
        if (this.handInEarly || this.studentExam?.submitted) {
```

Comment (sascha11110) [reply src/main/webapp/app/exam/participate/exam-participation.component.ts]: Tested it and it prevents the NPE.

**Comment by sascha11110** [comment]
@tiffca Thanks for testing! Yeah, this is intended. The students get a 3 minute grace period to finish the submission.

**Comment by schultek** [comment]
I noticed that unsaved changes get lost when handing in early. Is this intended or should it also trigger a save?

**Comment by sascha11110** [comment]
> I noticed that unsaved changes get lost when handing in early. Is this intended or should it also trigger a save?
> This happend also when clicking Continue.

@schultek  I've fixed this [here](https://github.com/ls1intum/Artemis/pull/1884/commits/1ec24d5fe03b05c441d3ff408a4fc5d6d831bfaa) and [here](https://github.com/ls1intum/Artemis/pull/1884/commits/cce981ed16a42347e62b4917a472e7e2567f1f2e). Maybe you can test again with the newest version.

**Comment by schultek** [comment]
Ok will do.

**Comment by schultek** [comment]
Probably not related to this PR, but what should happen if the student stays on the exam end screen until the timer runs out? For me it just stays there without any notification and i cannot click Finish anymore.

**Comment by sascha11110** [comment]
> Probably not related to this PR, but what should happen if the student stays on the exam end screen until the timer runs out? For me it just stays there without any notification and i cannot click Finish anymore.

@schultek Definitely not related to this PR :-) But a good catch... I will discuss it with Stephan.

@jpbernius I've fixed the problem with the unsaved submission.

**Comment by sascha11110** [comment]
> Tested again with the following results:
> When Clicking 'Hand in Early' and then 'Continue', previously unsaved changes are saved 👍
> When Clicking 'Hand in Early' and 'Finish', previously unsaved changes are submitted and shown in summary 👍
> When handed in Early, the Summary for Text Exercises shows a "View Result" button, which navigates to the exercise text editor, where one can submit changes again via the 'Submit' button 👎
> Same for Modeling Exercises 👎

Thanks for testing again! The two last points are related to a known bug on develop. This will be fixed in a hotfix by @sjagla27

**Pull request merged**